### PR TITLE
standardise on `liftF` and add `liftK` to transformers

### PR DESCRIPTION
**Pull request by SystemFw**
This PR deprecates `lift` on all transformers in favour of `liftF`, and adds a FunctionK version named `liftK` (for use with `mapK`).

I'm leaving a possible scalafix to a separate PR.

### CODE REVIEW

File: core/src/main/scala/cats/data/IndexedReaderWriterStateT.scala [hunk 1]
```diff
@@ -266,6 +266,25 @@ private[data] sealed trait CommonIRWSTConstructors {
...
+   * scala> b.value.runEmpty(true).value
+   * res0: (List[String], String, Option[Int]) = (List(),"",Some(1))
+   * }}}
+   */
+  def liftK[F[_], E, L, S](implicit F: Applicative[F], L: Monoid[L]): F ~> IndexedReaderWriterStateT[F, E, L, S, S, ?] =
+    new (F ~> IndexedReaderWriterStateT[F, E, L, S, S, ?]) {
+      def apply[A](fa: F[A]): IndexedReaderWriterStateT[F, E, L, S, S, A] = IndexedReaderWriterStateT.liftF(fa)
+    }
+
+  @deprecated("Use liftF instead", "1.0.0-RC2")
```
Comment (kailuowang) [inline core/src/main/scala/cats/data/IndexedReaderWriterStateT.scala]: There probably won't be a rc2, the next one is 1.0

Comment (SystemFw) [reply core/src/main/scala/cats/data/IndexedReaderWriterStateT.scala]: Yeah, wasn't quite sure. I'll fix tomorrow

**Comment by SystemFw** [comment]
Fixed the `RC` deprecation warning. I'm adding the scalafix now (changed my mind, I'll do it in this PR)

**Comment by kailuowang** [comment]
also need some mima exclusion
> [0m[[31merror[0m] [0m   filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("cats.data.CommonIRWSTConstructors.liftF")[0m
[0m[[31merror[0m] [0m * method liftK(cats.Applicative,cats.kernel.Monoid)cats.arrow.FunctionK in trait cats.data.CommonIRWSTConstructors is present only in current version[0m
[0m[[31merror[0m] [0m   filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("cats.data.CommonIRWSTConstructors.liftK")[0m
[0m[[31merror[0m] [0m * method liftF(java.lang.Object)cats.data.Kleisli in trait cats.data.KleisliFunctions is present only in current version[0m
[0m[[31merror[0m] [0m   filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("cats.data.KleisliFunctions.liftF")[0m
[0m[[31merror[0m] [0m * method liftK()cats.arrow.FunctionK in trait cats.data.KleisliFunctions is present only in current version[0m
[0m[[31merror[0m] [0m   filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("cats.data.KleisliFunctions.liftK")[0m
[0m[[31merror[0m] [0m * method liftF(java.lang.Object,cats.Applicative)cats.data.IndexedStateT in trait cats.data.CommonStateTConstructors is present only in current version[0m
[0m[[31merror[0m] [0m   filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("cats.data.CommonStateTConstructors.liftF")[0m
[0m[[31merror[0m] [0m * method liftK(cats.Applicative)cats.arrow.FunctionK in trait cats.data.CommonStateTConstructors is present only in current version[0m
[0m[[31merror[0m] [0m   filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("cats.data.CommonStateTConstructors.liftK")[0m

**Comment by SystemFw** [comment]
@kailuowang Any quick reference to how you exclude something from MiMa or should I RTFM? :)

**Comment by LukaJCB** [comment]
@SystemFw You can add them right here: https://github.com/typelevel/cats/blob/master/build.sbt#L212 :)

**Comment by SystemFw** [comment]
Thanks a lot! @LukaJCB

**Comment by aeons** [comment]
This is great. I've had to write the `OptionT` variant several times lately.

Is there a reason you're not using the kind-projector syntax?

```scala
def liftK[F[_]](implicit F: Functor[F]): F ~> OptionT[F, ?] = 
  λ[F ~> OptionT[F, ?]](OptionT.liftF(_))
```
(or `Lambda` instead of `λ`)

**Comment by SystemFw** [comment]
@aeons Yes, http4s was the initial use case :)

There's no reason for not using it, I might change it as I fix the other things

**Comment by SystemFw** [comment]
I think I'm mostly done, except my scalafix test doesn't compile ( it's telling me that `RWST` is not a member of `cats.data` ).
This happens because the scalafix is running on 0.9.0, and RWST was introduced later. How should I proceed?

P.S. the error in the travis seems spurious

**Comment by kailuowang** [comment]
the migration is for migrating from 0.9. ideally we need a new set of migration that migrates from 1.0 RC1, although I am not sure a simple rename in RWST justify that. I propose we just remove RWST from the test.

**Comment by SystemFw** [comment]
I think the errors are spurious and I should be done. The scalafix has fixes for both 0.9 and 1.0RC, the RC ones are not tested as discussed above

**Comment by SystemFw** [comment]
Some of the builds fail with 

> Could not find any member to link for "mapK"

should I address this?

**Comment by LukaJCB** [comment]
That seems odd, seems like an error during scaladoc compilation. Maybe someone else has an idea?

**Comment by SystemFw** [comment]
Yeah, I reference `[[mapK]]` in a scaladoc, might be pretty stupid. Can't really test locally since MiMa fails before on my machine ("versions not set" so it's my setup, it succeeds on Travis)

**Comment by LukaJCB** [comment]
I think it might be because `mapK` is on the instance while `liftF` is on the companion object, so they need to be referred to differently. Not sure how exactly though.

**Comment by SystemFw** [comment]
I think we agree on what the problem is then :)
I also think we agree in not knowing what the solution is :P
I'll investigate (this sort of search and replace PRs always end up taking ages for me)

**Comment by kailuowang** [comment]
Linking reference in scaladoc has always been problematic, the solutions people have been doing in the past, and I propose doing here, is to just remove the link.

**Comment by SystemFw** [comment]
so just `mapK`  rather than [[[mapK]]`? Sounds good to me :)

**Comment by SystemFw** [comment]
I need to fix the doctest for kleisli. It's the only bit of codecov that actually spotted a problem. Doing so right now

**Comment by frroliveira** [comment]
If you're still interested in getting the scaladoc references working, you can prefix `mapK` with the class name, e.g. `EitherT.mapK`. That's what I used in `IorT.liftK`

**Comment by kailuowang** [comment]
Close and reopen to trigger build

**Pull request merged**